Move from kotlinx.dom to Jackson for pom parsing

jpenilla opened the pull request:
Closes #4

jpenilla committed:
Use Jackson instead of kotlinx.dom for XML parsing

jpenilla committed:
Object map poms

jpenilla committed:
Reorganize

jpenilla committed:
Read maven plugin artifactIds

Proximyst commented on src/main/kotlin/xyz/jpenilla/toothpick/ConfigureSubprojects.kt:
```diff
@@ -120,41 +118,14 @@ private fun Project.configureServerProject() {
-        }
+    // Import relocations from server pom
+    val pom = subproject.pom ?: return@getting
+    val shadePlugin: ShadePlugin = pom.build.plugins.find { it is ShadePlugin } as ShadePlugin? ?: error("Could not find shade plugin in server pom!")
```
```suggestion
    val shadePlugin: ShadePlugin = pom.build.plugins.filterIsInstance().firstOrNull() ?: error("Could not find shade plugin in server pom!")
```

Proximyst commented on src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt:
```diff
@@ -0,0 +1,75 @@
+  val url: String
+)
+
+internal data class Dependency(
```
This does not fulfill the entirety of the definition of dependencies. Consider ignoring unknown properties or filling it: https://maven.apache.org/pom.html#Dependencies

Proximyst commented on src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt:
```diff
@@ -0,0 +1,75 @@
+  val dependencies: List<Dependency> = emptyList()
+)
+
+internal data class Repository(
```
This does not fulfill the entirety of the definition of repositories. Consider ignoring unknown properties or filling it: https://maven.apache.org/pom.html#Repositories

Proximyst commented on src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:
```diff
@@ -0,0 +1,59 @@
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+
+internal data class ShadePlugin(
```
This does not fulfill the entirety of the shade plugin parameters. Consider ignoring unknown parameters or filling it: https://github.com/apache/maven-shade-plugin/blob/master/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java#L99-L375

jpenilla replied:
Seems like it inherits `@JsonIgnoreProperties(ignoreUnknown = true)` from it's superclass

Proximyst replied:
Jackson does not inherit annotations, sadly. It must be explicitly defined.

Proximyst commented on src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:
```diff
@@ -0,0 +1,59 @@
+  val configuration: ShadeConfiguration = ShadeConfiguration()
+)
+
+internal data class Filter(
```
This does not fulfill the entirety of the filter parameters. Consider ignoring unknown properties or filling it: https://github.com/apache/maven-shade-plugin/blob/master/src/main/java/org/apache/maven/plugins/shade/mojo/ArchiveFilter.java#L29-L35

Proximyst reviewed (changes requested):
Looks good :)

jpenilla committed:
Suggested changes

jpenilla committed:
EXCLUDES

jpenilla committed:
Explicitly set ShadePlugin `@JsonIgnoreProperties`

jpenilla committed:
Also read relocation includes

jpenilla committed:
Import external Javadoc links from poms, link to JDK Javadocs

jpenilla committed:
Build updates

Also bump Shadow to 7.0.0

jpenilla committed:
Update GitHub Actions workflow

jpenilla committed:
Update Gradle to 7.0.2, Indra to 2.0.4, polyglot-version-catalogs to 5.0.0

jpenilla committed:
Use JavaExec for both run tasks

jpenilla committed:
Bump Configurate to 4.1.1

jpenilla committed:
- Clean up usage of rootProject
- Apply `net.kyori.indra.git` plugin by default
- Add companion settings plugin to simplify setting up a toothpick project
- Disable jar artifacts on parent project by default
- Reorder initialization so that we can use the generated extension accessor instead of our own extension method
- Bumped version to 1.1.0-SNAPSHOT

jpenilla committed:
Add property for debugging importMCDev

jpenilla committed:
Fix importMCDev debug

jpenilla committed:
Register tasks on plugin apply for accessor generation

kennytv commented:
thank you jomp, very cool
